dev/src/types.ts

```diff
@@ -32,10 +33,57 @@ export type ApiMapValue = {
   [k: string]: google.firestore.v1beta1.IValue
 };
 
-/** JavaScript input from the API layer. */
+/**
+ * @private
+ * JavaScript input from the API layer.
+ */
 // tslint:disable-next-line:no-any
 export type UserInput = any;
 
+// tslint:disable-next-line:no-any
+export type GapicClient = any;
+
+/**
+ * Settings used to directly configure a `Firestore` instance.
+ */
+export interface Settings {
+  /**
+   * The Firestore Project ID. Can be omitted in environments that support
+   * `Application Default Credentials` {@see https://cloud.google.com/docs/authentication}
+   */
+  projectId?: string;
+
+  /**
+   * Local file containing the Service Account credentials. Can be omitted
+   * in environments that support `Application Default Credentials`
+   * {@see https://cloud.google.com/docs/authentication}
+   */
+  keyFilename?: string;
+
+  /**
+   * Enables the use of `Timestamp`s for timestamp fields in
+   * `DocumentSnapshot`s.
+   *
+   * Currently, Firestore returns timestamp fields as `Date` but `Date` only
+   * supports millisecond precision, which leads to truncation and causes
+   * unexpected behavior when using a timestamp from a snapshot as a part
+   * of a subsequent query.
+   *
+   * Setting `timestampsInSnapshots` to true will cause Firestore to return
+   * `Timestamp` values instead of `Date` avoiding this kind of problem. To
+   * make this work you must also change any code that uses `Date` to use
+   * `Timestamp` instead.
+   *
+   * NOTE: in the future `timestampsInSnapshots: true` will become the
+   * default and this option will be removed so you should change your code to
+   * use `Timestamp` now and opt-in to this new behavior as soon as you can.
+   */
+  timestampsInSnapshots?: boolean;
+
+  // tslint:disable-next-line:no-any
+  [key: string]: any;  // Accept other properties, such as GRPC settings.
```

Comment:
By the way, what version of TypeScript are we using?  TypeScript 3.0 introduces the type-safe type `unknown`, which may be more suitable than `any` for some of these things...

Reply:
We are using TS 3 as of last week. I filed a bug for `unknown` (b/117464450).